src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph_RemoveProject.cs

```diff
@@ -0,0 +1,167 @@
+                // references in each of those cases directly.
+                foreach (var id in referencingProjects)
+                {
+                    builder[id] = builder[id].Remove(projectId);
+                }
+            }
+            else
+            {
+                // We don't know which projects reference 'projectId', so iterate over all known projects and remove
+                // 'projectId' from the set of references if it exists.
+                foreach (var (id, references) in referencesMap)
+                {
+                    builder[id] = references.Remove(projectId);
+                }
+            }
+
+            // Finally, remove 'projectId' itself.
+            builder.Remove(projectId);
+            return builder.ToImmutable();
+        }
+
+        /// <summary>
+        /// Computes a new <see cref="_lazyReverseReferencesMap"/> for the removal of a project.
+        /// </summary>
+        private static ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>>? ComputeNewReverseReferencesMapForRemovedProject(
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> existingForwardReferencesMap,
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>>? existingReverseReferencesMap,
+            ProjectId projectId)
+        {
+            if (existingReverseReferencesMap is null)
+            {
+                // The map was never calculated for the previous graph, so there is nothing to update.
+                return null;
+            }
+
+            var builder = existingReverseReferencesMap.ToBuilder();
+
+            // Iterate over each project referenced by 'projectId', which is now being removed. Update the reverse
+            // references map for the project to no longer include 'projectId' in the list.
+            if (existingForwardReferencesMap.TryGetValue(projectId, out var referencedProjectIds))
+            {
+                foreach (var referencedProjectId in referencedProjectIds)
+                {
+                    if (builder.TryGetValue(referencedProjectId, out var reverseDependencies))
+                    {
+                        builder[referencedProjectId] = reverseDependencies.Remove(projectId);
+                    }
+                }
+            }
+
+            // Finally, remove 'projectId' itself.
+            builder.Remove(projectId);
+            return builder.ToImmutable();
+        }
+
+        /// <summary>
+        /// Computes a new <see cref="_transitiveReferencesMap"/> for the removal of a project.
+        /// </summary>
+        /// <seealso cref="ComputeNewReverseTransitiveReferencesMapForRemovedProject"/>
+        private static ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> ComputeNewTransitiveReferencesMapForRemovedProject(
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> existingTransitiveReferencesMap,
+            ProjectId projectId)
```

Comment:
```suggestion
            ProjectId removedProjectId)
```

Might be a bit easier than having both "project" and "projectId" in the code below. Same comment could apply to rest of the methods in this file.